### DISCUSSION TITLE
Create and checkout a branch?

### DISCUSSION THREAD

**Discussion by dbanty** [GitoxideLabs/gitoxide#879] [Q&A] [answered]
I'm trying to finish replacing my usage of `git2` and the only thing I can't figure out is how to create and checkout a branch. Is this possible yet using `gix`?

For reference, here are some of the relevant functions in `git2`:

- [Repository::branch](https://docs.rs/git2/0.17.2/git2/struct.Repository.html#method.branch)
- [Repository::checkout_head](https://docs.rs/git2/0.17.2/git2/struct.Repository.html#method.checkout_head)

**Comment by Byron** [comment, marked as answer]
Branch adjustments are [fully supported](https://docs.rs/gix/0.44.1/gix/struct.Repository.html#method.reference), but handing the working tree is not. Thus, checking out a branch isn't possible yet. It should be possible to follow progress [in this tracking issue](https://github.com/Byron/gitoxide/issues/301)

If the usage of `git2` is for setting up test fixtures, `gix-testtools` could offload all this work into handy scripts that use `git` instead with various built-in features to make this as fast as possible.

**Reply by Ltrlg** [reply]
I guess you meant #301?

**Reply by Byron** [reply]
Yes, that's the one. Fixed it above.

**Reply by dbanty** [reply]
Thanks! I’ll follow along excitedly 😄

**Comment by PoignardAzur** [comment]
> Branch adjustments are [fully supported](https://docs.rs/gix/0.44.1/gix/struct.Repository.html#method.reference), but handing the working tree is not.

To be clear, does that mean doing a `git add` with gix isn't possible? (Looking at the linked issue, it sounds like the answer's yes)

**Reply by Byron** [reply]
It would be possible to implement `git add` using plumbing crates if the `index-to-tree` (or `tree-from-index`) transformation would be implemented. Even then it would be quite cumbersome to do it, so I still think the answer is: Yes, `git add` isn't implemented in `gitoxide`.
